src/main/java/net/snowflake/client/jdbc/SnowflakeStatementV1.java

```diff
@@ -214,14 +214,16 @@ long executeUpdateInternal(
       sfResultSet.setSession(this.connection.getSFBaseSession());
       updateCount = ResultUtil.calculateUpdateCount(sfResultSet);
       queryID = sfResultSet.getQueryId();
-      statementMetaData =
-          new SFStatementMetaData(
-              sfResultSet.getMetaData(),
-              sfResultSet.getStatementType(),
-              sfResultSet.getNumberOfBinds(),
-              sfResultSet.isArrayBindSupported(),
-              sfResultSet.getMetaDataOfBinds(),
-              true); // valid metadata
+      if (!statementMetaData.isValidMetaData()) {
```

Comment:
Have you tried executing two different SQL statements using same JDBC stmt object? e.g. stmt.exeuteQuery("SELECT * from t1") and stmt.executeQuery("SELECT * from t2"); Display metadata for both query and make sure that's valid. My guess is still we need some more fix here. Also in your test make sure both tables has different sets of columns.

Reply:
The name made it confusing. I refactored the code so that it's clear the statement metadata is only for preparedStatement, not Statement class.  The Statement class does not access this metadata and the PreparedStatement class is forced to create a new PreparedStatement object every time the sql query changes. Therefore, we cannot face the issue of the metadata being repeated for different sql queries.